### PR TITLE
Cherry pick KOGITO-5603, KOGITO-3793

### PR DESCRIPTION
**Pull request by vaibhavjainwiz**
Many thanks for submitting your Pull Request :heart:! 

https://issues.redhat.com/browse/KOGITO-5603
https://issues.redhat.com/browse/KOGITO-3793

Please make sure your PR meets the following requirements:

- [x] You have read the [contributors' guide](https://github.com/kiegroup/kogito-operator/blob/main/README.md#contributing-to-the-kogito-operator)
- [x] Pull Request title is properly formatted: `[KOGITO-XYZ] Subject`
- [x] Pull Request contains a link to the JIRA issue
- [x] Pull Request contains a description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Your feature/bug fix has a unit test that verifies it
- [x] You've tested the new feature/bug fix in an actual OpenShift cluster
- [ ] You've added a [RELEASE_NOTES.md](RELEASE_NOTES.md) entry regarding this change

### CODE REVIEW

File: api/v1/kogitoinfra_types.go [hunk 1]
```diff
@@ -306,13 +364,17 @@ func (k KogitoInfraVolume) GetNamedVolume() api.ConfigVolumeInterface {
 // +kubebuilder:printcolumn:name="API Version",type="string",JSONPath=".spec.resource.apiVersion",description="Kubernetes CR API Version"
 // +kubebuilder:printcolumn:name="Status",type="string",JSONPath=".status.condition.status",description="General Status of this resource bind"
 // +kubebuilder:printcolumn:name="Reason",type="string",JSONPath=".status.condition.reason",description="Status reason"
-// +operator-sdk:gen-csv:customresourcedefinitions.displayName="Kogito Infra"
-// +operator-sdk:gen-csv:customresourcedefinitions.resources="Kafka,kafka.strimzi.io/v1beta2,\"A Kafka instance\""
-// +operator-sdk:gen-csv:customresourcedefinitions.resources="Secret,v1,\"A Kubernetes Secret\""
+// +operator-sdk:csv:customresourcedefinitions:displayName="Kogito Infra"
+// +operator-sdk:csv:customresourcedefinitions:resources={{Kafka,kafka.strimzi.io/v1beta2,"A Kafka instance"}}
+// +operator-sdk:csv:customresourcedefinitions:resources={{Infinispan,infinispan.org/v1,"A Infinispan instance"}}
+// +operator-sdk:csv:customresourcedefinitions:resources={{Keycloak,keycloak.org/v1alpha1,"A Keycloak Instance"}}
```
Comment (sutaakar) [inline api/v1/kogitoinfra_types.go:370]:
```suggestion
```

File: controllers/kogitoruntime_controller.go [hunk 2]
```diff
@@ -118,6 +122,13 @@ func (r *KogitoRuntimeReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		For(&rhpamv1.KogitoRuntime{}, builder.WithPredicates(pred)).
 		Owns(&corev1.Service{}).Owns(&appsv1.Deployment{}).Owns(&corev1.ConfigMap{})
 
+	infraHandler := &handler.EnqueueRequestForOwner{IsController: false, OwnerType: &v1beta1.KogitoRuntime{}}
```
Comment (sutaakar) [inline controllers/kogitoruntime_controller.go:125]:
```suggestion
	infraHandler := &handler.EnqueueRequestForOwner{IsController: false, OwnerType: &rhpamv1.KogitoRuntime{}}
```

File: controllers/kogitoruntime_controller.go [hunk 3]
```diff
@@ -118,6 +122,13 @@ func (r *KogitoRuntimeReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+	infraPred := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return reflect.DeepEqual(e.MetaNew.GetOwnerReferences(), e.MetaOld.GetOwnerReferences())
+		},
+	}
+	b.Watches(&source.Kind{Type: &v1beta1.KogitoInfra{}}, infraHandler, builder.WithPredicates(infraPred))
```
Comment (sutaakar) [inline controllers/kogitoruntime_controller.go:131]:
```suggestion
	b.Watches(&source.Kind{Type: &rhpamv1.KogitoInfra{}}, infraHandler, builder.WithPredicates(infraPred))
```

File: api/v1/kogitoinfra_types.go (hunk 1 above)
Comment (vaibhavjainwiz) [reply api/v1/kogitoinfra_types.go:370]: with Kogito-infra integration in RHPAM kogito operator now we support them all

File: controllers/kogitoinfra_controller.go [hunk 4]
```diff
@@ -43,9 +43,16 @@ type KogitoInfraReconciler struct {
...
+// +kubebuilder:rbac:groups=apps,resources=deployments;replicasets,verbs=get;create;list;watch;create;delete;update
+// +kubebuilder:rbac:groups=infinispan.org,resources=infinispans,verbs=get;create;list;delete;watch
 // +kubebuilder:rbac:groups=kafka.strimzi.io,resources=kafkas;kafkatopics,verbs=get;create;list;delete;watch
+// +kubebuilder:rbac:groups=keycloak.org,resources=keycloaks,verbs=get;create;list;delete;watch
 // +kubebuilder:rbac:groups=apps,resources=deployments/finalizers,verbs=update
+// +kubebuilder:rbac:groups=eventing.knative.dev,resources=brokers,verbs=get;list;watch
+// +kubebuilder:rbac:groups=eventing.knative.dev,resources=triggers,verbs=get;list;watch;create;delete;update
+// +kubebuilder:rbac:groups=sources.knative.dev,resources=sinkbindings,verbs=get;list;watch;create;delete;update
 // +kubebuilder:rbac:groups=integreatly.org,resources=grafanadashboards,verbs=get;create;list;watch;create;delete;update
+// +kubebuilder:rbac:groups=mongodb.com,resources=mongodb,verbs=get;create;list;watch;delete
```
[the remark this answers is not in the document]
Comment (vaibhavjainwiz) [reply controllers/kogitoinfra_controller.go:55]: with Kogito-infra integration in RHPAM kogito operator now we support them all

File: controllers/kogitoruntime_controller.go (hunk 2 above)
Comment (vaibhavjainwiz) [reply controllers/kogitoruntime_controller.go]: done

File: controllers/kogitoruntime_controller.go (hunk 3 above)
Comment (vaibhavjainwiz) [reply controllers/kogitoruntime_controller.go]: done

**Comment by sutaakar** [comment]
@vaibhavjainwiz Creating KogitoInfra is causing operator crash:
```
apiVersion: rhpam.kiegroup.org/v1
kind: KogitoInfra
metadata:
  name: kogito-kafka-infra
spec:
  envs:
    - name: AAA
      value: BBB
  infraProperties:
    qqqq: asdfds
```
error:
```
{"level":"error","ts":1630485239.993891,"logger":"kogitoInfra-controller","msg":"Error occurs while fetching deployed kogito infra instance","name":"kogito-kafka-infra","namespace":"ksuta-test","error":"no kind is registered for the type v1beta1.KogitoInfra in scheme \"go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/runtime/scheme.go:100\"","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/opt/app-root/src/go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132\ngithub.com/kiegroup/kogito-operator/internal.(*kogitoInfraHandler).FetchKogitoInfraInstance\n\t/opt/app-root/src/go/pkg/mod/github.com/kiegroup/kogito-operator@v0.12.1-0.20210827160405-f6a73130fe0b/internal/kogitoinfra.go:55\ngithub.com/kiegroup/kogito-operator/core/kogitoinfra.(*statusHandler).isStatusChanged\n\t/opt/app-root/src/go/pkg/mod/github.com/kiegroup/kogito-operator@v0.12.1-0.20210827160405-f6a73130fe0b/core/kogitoinfra/status.go:70\ngithub.com/kiegroup/kogito-operator/core/kogitoinfra.(*statusHandler).UpdateBaseStatus\n\t/opt/app-root/src/go/pkg/mod/github.com/kiegroup/kogito-operator@v0.12.1-0.20210827160405-f6a73130fe0b/core/kogitoinfra/status.go:59\ngithub.com/kiegroup/rhpam-kogito-operator/controllers.(*KogitoInfraReconciler).Reconcile\n\t/workspace/controllers/kogitoinfra_controller.go:125\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.3/pkg/internal/controller/controller.go:244\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.3/pkg/internal/controller/controller.go:218\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.3/pkg/internal/controller/controller.go:197\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/opt/app-root/src/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/opt/app-root/src/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/opt/app-root/src/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/opt/app-root/src/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:90"}
```

**Review by ricardozanini — APPROVED** [review]

**Comment by vaibhavjainwiz** [comment]
> @vaibhavjainwiz Creating KogitoInfra is causing operator crash:
> 
> ```
> apiVersion: rhpam.kiegroup.org/v1
> kind: KogitoInfra
> metadata:
>   name: kogito-kafka-infra
> spec:
>   envs:
>     - name: AAA
>       value: BBB
>   infraProperties:
>     qqqq: asdfds
> ```
> 
> error:
> 
> ```
> {"level":"error","ts":1630485239.993891,"logger":"kogitoInfra-controller","msg":"Error occurs while fetching deployed kogito infra instance","name":"kogito-kafka-infra","namespace":"ksuta-test","error":"no kind is registered for the type v1beta1.KogitoInfra in scheme \"go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/runtime/scheme.go:100\"","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/opt/app-root/src/go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132\ngithub.com/kiegroup/kogito-operator/internal.(*kogitoInfraHandler).FetchKogitoInfraInstance\n\t/opt/app-root/src/go/pkg/mod/github.com/kiegroup/kogito-operator@v0.12.1-0.20210827160405-f6a73130fe0b/internal/kogitoinfra.go:55\ngithub.com/kiegroup/kogito-operator/core/kogitoinfra.(*statusHandler).isStatusChanged\n\t/opt/app-root/src/go/pkg/mod/github.com/kiegroup/kogito-operator@v0.12.1-0.20210827160405-f6a73130fe0b/core/kogitoinfra/status.go:70\ngithub.com/kiegroup/kogito-operator/core/kogitoinfra.(*statusHandler).UpdateBaseStatus\n\t/opt/app-root/src/go/pkg/mod/github.com/kiegroup/kogito-operator@v0.12.1-0.20210827160405-f6a73130fe0b/core/kogitoinfra/status.go:59\ngithub.com/kiegroup/rhpam-kogito-operator/controllers.(*KogitoInfraReconciler).Reconcile\n\t/workspace/controllers/kogitoinfra_controller.go:125\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.3/pkg/internal/controller/controller.go:244\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.3/pkg/internal/controller/controller.go:218\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.3/pkg/internal/controller/controller.go:197\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/opt/app-root/src/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/opt/app-root/src/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/opt/app-root/src/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/opt/app-root/src/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:90"}
> ```

This issue would be fixed with https://github.com/kiegroup/kogito-operator/pull/996

**Review by sutaakar — APPROVED** [review]
Works as expected

**Comment by spolti** [comment]
cherry-picks https://github.com/kiegroup/kogito-operator/pull/967
cherry-picks https://github.com/kiegroup/kogito-operator/pull/988

**Review by spolti — APPROVED** [review]

**Pull request merged**